src/librustc_typeck/check/expr.rs

```diff
@@ -1405,6 +1402,20 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 _ => {}
             }
 
+            let deref_t = match expr_t.kind {
+                ty::Ref(_, ref_t, _) => ref_t,
+                _ => &expr_t
+            };
+            match deref_t.kind {
```

Comment:
```suggestion
            match expr_t.peel_refs().kind {
```
(under the assumption that we want to peel off as many references as possible)

Reply:
Peeling off all the references is consistent with `expr_t.is_primitive_ty()` which got us into this branch (a few lines above), so I suspect we do. Will change.